### PR TITLE
Enable Simplicity script tests

### PR DESCRIPTION
**Pull request by uncomputable**
This PR enables the asset tests and extends them to support Simplicity.

The first two commits are cherry-picked from https://github.com/ElementsProject/elements/pull/1232/commits

498ed9f396f6a95741c858816448e6598e7e5231 extends the format of [script_assets_test.json](https://github.com/ElementsProject/qa-assets/blob/master/unit_test_data/script_assets_test.json) to include error codes:

Before:
```
"failure":{"scriptSig":"","witness":[]}
```

After:
```
"failure":{"scriptSig":"","witness":[],"error":"SIMPLICITY_BITSTREAM_UNUSED_BYTES"}
```

The error in the JSON is the expected error, which is compared against the actual error that `VerifyScript` returns. The test fails if they are not equal.

If the JSON does not include an expected error, then this comparison is skipped. The test behaves as normal.

### CODE REVIEW

**Comment by uncomputable** [comment]
The linter seems to be confused by Simplicity file names. This is something that the base branch must fix.
```
File 'src/simplicity/ctx8Pruned.h' does not not match the allowed source filename regexp ('^[a-z0-9_./-]+$'), or the exception regexp (^src/(secp256k1/|univalue/|test/fuzz/FuzzedDataProvider.h)).
File 'src/simplicity/ctx8Unpruned.h' does not not match the allowed source filename regexp ('^[a-z0-9_./-]+$'), or the exception regexp (^src/(secp256k1/|univalue/|test/fuzz/FuzzedDataProvider.h)).
File 'src/simplicity/hashBlock.h' does not not match the allowed source filename regexp ('^[a-z0-9_./-]+$'), or the exception regexp (^src/(secp256k1/|univalue/|test/fuzz/FuzzedDataProvider.h)).
File 'src/simplicity/include/simplicity/errorCodes.h' does not not match the allowed source filename regexp ('^[a-z0-9_./-]+$'), or the exception regexp (^src/(secp256k1/|univalue/|test/fuzz/FuzzedDataProvider.h)).
File 'src/simplicity/primitive/elements/checkSigHashAllTx1.h' does not not match the allowed source filename regexp ('^[a-z0-9_./-]+$'), or the exception regexp (^src/(secp256k1/|univalue/|test/fuzz/FuzzedDataProvider.h)).
File 'src/simplicity/typeInference.h' does not not match the allowed source filename regexp ('^[a-z0-9_./-]+$'), or the exception regexp (^src/(secp256k1/|univalue/|test/fuzz/FuzzedDataProvider.h)).
ERROR: There were 6 failed tests in the lint-files.py lint test. Please resolve the above errors.
```

IDK what the sanitizers detected here. The CI check is marked as [jammy]. Reviewers, please provide feedback on how to deal with this CI failure.
```
Running tests: allocator_tests from test/allocator_tests.cpp
FATAL: ThreadSanitizer: unexpected memory mapping 0x5b88add23000-0x5b88addba000
make[3]: *** [Makefile:21156: test/pegin_witness_tests.cpp.test] Error 1
make[3]: *** Waiting for unfinished jobs....
FATAL: ThreadSanitizer: unexpected memory mapping 0x5a90d23fd000-0x5a90d2494000
make[3]: *** [Makefile:21156: test/arith_uint256_tests.cpp.test] Error 1
FATAL: ThreadSanitizer: unexpected memory mapping 0x5b3683cba000-0x5b3683d51000
FATAL: ThreadSanitizer: unexpected memory mapping 0x5d550b976000-0x5d550ba0d000
```

**Comment by delta1** [comment]
> The linter seems to be confused by Simplicity file names. This is something that the base branch must fix. 

I'd suggest since this branch is introducing `src/simplicity` that it should also add that to the exception regexp in `lint-files.py`  

Don't worry about the TSan failure, it's unrelated to the simplicity branch.

**Comment by delta1** [comment]
> extends the format of [script_assets_test.json](https://github.com/ElementsProject/qa-assets/blob/master/unit_test_data/script_assets_test.json) to include error codes 

@uncomputable could you please put in a PR to the qa-assets repo with your updated json file?

**Comment by uncomputable** [comment]
Depends on #1281

**Comment by uncomputable** [comment]
Rebased

**Comment by uncomputable** [comment]
Rebased

File: src/test/script_tests.cpp [hunk 1]
```diff
@@ -101,6 +101,40 @@ static ScriptErrorDesc script_errors[]={
     {SCRIPT_ERR_WITNESS_PUBKEYTYPE, "WITNESS_PUBKEYTYPE"},
     {SCRIPT_ERR_OP_CODESEPARATOR, "OP_CODESEPARATOR"},
     {SCRIPT_ERR_SIG_FINDANDDELETE, "SIG_FINDANDDELETE"},
+    // Elements
+    {SCRIPT_ERR_ERROR_COUNT, "ERROR_COUNT"},
```
Comment (roconnor-blockstream) [inline src/test/script_tests.cpp:106]: As I mentioned in #1285, this isn't a real error, and it should probably be removed from this list.

Comment (uncomputable) [reply src/test/script_tests.cpp:106]: I can remove `SCRIPT_ERR_ERROR_COUNT` in my PR, but ultimately this has to be fixed on master. We could upstream my commit.

**Comment by uncomputable** [comment]
I rebased to remove `SCRIPT_ERR_ERROR_COUNT`

**Comment by uncomputable** [comment]
Rebased and ready for review @roconnor-blockstream

**Comment by roconnor-blockstream** [comment]
Do you have your own `script_assets_test.json` that I should be testing against?

**Comment by uncomputable** [comment]
I created a [gist with the latest Simplicity script assets](https://gist.github.com/uncomputable/1c52396bf4398d425471822da38e3433).

Feel free to also test against the [existing Elements script assets](https://github.com/ElementsProject/qa-assets/blob/master/unit_test_data/script_assets_test.json) to check backwards compatibility.

**Review by roconnor-blockstream — APPROVED** [review]
I'm not an expert on the BOOST test framework, but LGTM.

**Pull request merged**